token/metadata.go

```diff
@@ -48,3 +53,178 @@ func (m *Metadata) SpentTokenID() []*token2.ID {
 	}
 	return res
 }
+
+// FilterBy returns a new Metadata containing only the metadata that matches the given enrollment IDs.
+// For Issue actions: the returned metadata will contain only the outputs whose owner has the given enrollment IDs.
+// For Transfer actions:
+// - The returned metadata will contain only the outputs whose owner has the given enrollment IDs;
+// - The senders are all included;
+// - The list of token IDs will be empty.
+func (m *Metadata) FilterBy(eIDs ...string) (*Metadata, error) {
+	res := &Metadata{
+		tms:                  m.tms,
+		tokenRequestMetadata: &api2.TokenRequestMetadata{},
+	}
+
+	// filter issues
+	for _, issue := range m.tokenRequestMetadata.Issues {
+		// By default, an issue action contain only one receiver
+
+		// If the receiver has the given enrollment ID, add it
+		recipientEID, err := m.tms.GetEnrollmentID(issue.ReceiversAuditInfos[0])
+		if err != nil {
+			return nil, errors.Wrap(err, "failed getting enrollment ID")
+		}
+		if search(eIDs, recipientEID) == -1 {
+			logger.Debugf("skipping issue for [%s]", recipientEID)
+			continue
+		}
+		res.tokenRequestMetadata.Issues = append(res.tokenRequestMetadata.Issues, issue)
+	}
+
+	// filter transfers
+	for _, transfer := range m.tokenRequestMetadata.Transfers {
+		transferRes := api2.TransferMetadata{}
+
+		// Filter senders:
+		// All senders are appended
+		for i, sender := range transfer.Senders {
+			transferRes.Senders = append(transferRes.Senders, sender)
+			transferRes.SenderAuditInfos = append(transferRes.SenderAuditInfos, transfer.SenderAuditInfos[i])
+		}
+
+		// Filter outputs
+		// if the receiver has the given enrollment ID, add it. Otherwise, add empty entries
+		for i, auditInfo := range transfer.ReceiverAuditInfos {
+			recipientEID, err := m.tms.GetEnrollmentID(auditInfo)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed getting enrollment ID")
+			}
+			var Outputs []byte
+			var TokenInfo []byte
+			var Receivers view.Identity
+			var ReceiverIsSender bool
+			var ReceiverAuditInfos []byte
+
+			if search(eIDs, recipientEID) != -1 {
+				logger.Debugf("keeping transfer for [%s]", recipientEID)
+				Outputs = transfer.Outputs[i]
+				TokenInfo = transfer.TokenInfo[i]
+				Receivers = transfer.Receivers[i]
+				ReceiverIsSender = transfer.ReceiverIsSender[i]
+				ReceiverAuditInfos = transfer.ReceiverAuditInfos[i]
```

Comment:
can we get rid of this information? it's already know to the recipient with eid.

Reply:
it is true that it is known but the code assumes it will be there :(
Getting rid of it would require more effort :(

Reply:
OK - no problem